### PR TITLE
Added tests for changing a list size being iterated on (#65608)

### PR DESCRIPTION
**Pull request by SkiFoD**
If I understand everything right then the concern is that we don't have tests to cover the list size change while the list is being iterated on. 
So I added some simple tests to make sure that the "InvalidOperationExceptions" is thrown when the size is changing.
Please feel free to let me know if I missed something.

### CODE REVIEW

**Comment by SkiFoD** [comment]
@stephentoub cc

**Comment by stephentoub** [comment]
> If I understand everything right then the concern is that we don't have tests to cover the list size change while the list is being iterated on.

https://github.com/dotnet/runtime/issues/65608 is specifically about a few methods that iterate through the list using the indexer rather than enumerator and _don't_ check that the list has been modified between invocations of the user-provided delegate.  For example, this code:
```C#
var list = new List<int>() { 1, 2, 3 };
var results = list.FindAll(i =>
{
    if (i < 10)
    {
        list.Add(i + 1);
    }
    return true;
});
Console.WriteLine(string.Join(", ", list));
Console.WriteLine(string.Join(", ", results));
```
outputs:
```
1, 2, 3, 2, 3, 4, 3, 4, 5, 4, 5, 6, 5, 6, 7, 6, 7, 8, 7, 8, 9, 8, 9, 10, 9, 10, 10
1, 2, 3, 2, 3, 4, 3, 4, 5, 4, 5, 6, 5, 6, 7, 6, 7, 8, 7, 8, 9, 8, 9, 10, 9, 10, 10
```
We're lacking tests that validate this continues to work.  Had we merged https://github.com/dotnet/runtime/pull/65572, this would have changed behavior because it would have cached the size of the list at the beginning of the method invocation rather than re-reading the size on each iteration.

While I don't know if we have coverage for everything you've added in this PR, we definitely already have some of it.  For example, if I comment out the _version++ lines in List and then run our collection tests, there are 52 failures.

**Comment by danmoseley** [comment]
@stephentoub does this address your feedback?

File: src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Find.cs [hunk 1]
```diff
@@ -287,6 +287,23 @@ public void Find_VerifyDuplicates(int count)
...
+            {
+                if (i < 4)
+                {
+                    list.Add(i + 1);
+                }
+
+                return true;
+            });
+
+            Assert.Equal(list, results);
```
Comment (stephentoub) [inline src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Find.cs:304]: This should validate that list contains exactly the sequence we expect it to.

File: src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Find.cs [hunk 2]
```diff
@@ -981,6 +998,23 @@ public void FindAll_VerifyDuplicates(int count)
...
```
Comment (stephentoub) [inline src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Find.cs:1015]: This looks like an exact duplicate of the test added earlier?

Comment (SkiFoD) [reply src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Find.cs:1015]: Yeah, you are right. My bad.

File: src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Misc.cs [hunk 3]
```diff
@@ -1099,5 +1099,23 @@ public static void TrueForAllTests_Negative()
             intDriver.TrueForAll_VerifyExceptions(intArray);
             stringDriver.TrueForAll_VerifyExceptions(stringArray);
         }
+
+        [Fact]
+        public static void TrueForAll_ListSizeCanBeChanged()
+        {
+            var list = new List<int>() { 1, 2, 3 };
+            var results = new List<int> { 1, 2, 3, 2, 3, 4, 3, 4, 4 };
+            list.TrueForAll(i =>
```
[the remark this answers is not in the document]
Comment (SkiFoD) [reply src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Misc.cs:1108]: @stephentoub I'm sorry, but I don't get this one.  Do we want to make sure that the TrueForAll returns what we expect it to?

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Misc.cs:1108]: Yes, we should add an assert for the return value.

[the remark this answers is not in the document]
Comment (SkiFoD) [reply src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Misc.cs:1108]: Should it to be applied to the FindAll/Find tests that I added?

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Collections/tests/Generic/List/List.Generic.Tests.Misc.cs]: > Should it to be applied to the FindAll/Find tests that I added?

You're already validating the results of the those calls.  It was just this one that was missing an assert.

**Review by stephentoub — APPROVED** [review]
Thanks!

**Pull request merged**